### PR TITLE
chore: Type 3d-tile and I3S loaders.

### PR DESCRIPTION
**Pull request by ibgreen**
We should make a push towards getting 4.0 ready for release:
- One of the key checklist items for 4.0 is that all loaders should be typed. 
- This PR starts a pass on the 3d tiles and I3S loaders.

### CODE REVIEW

File: modules/3d-tiles/src/cesium-ion-loader.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ async function preload(url, options = {}) {
 /**
  * Loader for 3D tiles from Cesium ION
  */
-export const CesiumIonLoader: LoaderWithParser = {
+export const CesiumIonLoader: LoaderWithParser<unknown, never, LoaderOptions> = {
```
Comment (ibgreen) [inline modules/3d-tiles/src/cesium-ion-loader.ts:21]: We should define (and export) a `CesiumIONLoaderOptions` and clear up the options mess below and make sure types are observed in the code.

File: modules/3d-tiles/src/tile-3d-subtree-loader.ts [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+import type {Subtree} from './types';
 import parse3DTilesSubtree from './lib/parsers/helpers/parse-3d-tile-subtree';
 import {VERSION} from './lib/utils/version';
 
 /**
  * Loader for 3D Tiles Subtree
- *
  */
-export const Tile3DSubtreeLoader: LoaderWithParser = {
+export const Tile3DSubtreeLoader: LoaderWithParser<Subtree, never, LoaderOptions> = {
```
Comment (ibgreen) [inline modules/3d-tiles/src/tile-3d-subtree-loader.ts:9]: - Return types of the loaders should be exported from the module on the top level. 
- We should probably rename this to `Tiles3DSubtree` if we make it public.
-  We should define and export a `Tiles3DLoaderOptions` even if we don't have any unique options now. That lets us add additional options later in a non-breaking way.

File: modules/3d-tiles/src/tile-3d-writer.ts [hunk 3]
```diff
@@ -1,11 +1,11 @@
-import type {Writer} from '@loaders.gl/loader-utils';
+import type {Writer, WriterOptions} from '@loaders.gl/loader-utils';
 import {VERSION} from './lib/utils/version';
 import encode3DTile from './lib/encoders/encode-3d-tile';
 
 /**
  * Exporter for 3D Tiles
  */
-export const Tile3DWriter: Writer = {
+export const Tile3DWriter: Writer<unknown, never, WriterOptions> = {
```
Comment (ibgreen) [inline modules/3d-tiles/src/tile-3d-writer.ts:8]: - The input data to this Writer should have a type that is also exported from the module.
- We should have a matching WriterOptions type even if it doesn't add anything.

File: modules/3d-tiles/src/tiles-3d-loader.ts [hunk 4]
```diff
@@ -29,7 +29,7 @@ export type Tiles3DLoaderOptions = LoaderOptions &
 /**
  * Loader for 3D Tiles
  */
-export const Tiles3DLoader: LoaderWithParser = {
+export const Tiles3DLoader: LoaderWithParser<unknown, never, Tiles3DLoaderOptions> = {
```
Comment (ibgreen) [inline modules/3d-tiles/src/tiles-3d-loader.ts:32]: - Typing the returned value from this loader is the single biggest service we can do to the user. Returning `any` or `unknown` is really not helping anyone.
- I would say that this is a P0 for loaders.gl v4

File: modules/3d-tiles/src/tz3-loader.ts [hunk 5]
```diff
@@ -17,7 +17,7 @@ export type TZ3LoaderOptions = LoaderOptions & {
 /**
  * Loader for 3tz packages
  */
-export const TZ3Loader: LoaderWithParser<Buffer, never, TZ3LoaderOptions> = {
+export const TZ3Loader: LoaderWithParser<ArrayBuffer, never, TZ3LoaderOptions> = {
```
Comment (ibgreen) [inline modules/3d-tiles/src/tz3-loader.ts:20]: I am very confused about the `Buffer` type. It is named similar to the Node.js type, but is actually our own type and seems to do the same thing as a typed array or ArrayBuffer. Let's see if we can avoid it, if not give it a more distinctive name,

File: modules/i3s/src/arcgis-webscene-loader.ts [hunk 6]
```diff
@@ -11,7 +11,7 @@ const VERSION = typeof __VERSION__ !== 'undefined' ? __VERSION__ : 'beta';
  * Loader for ArcGis WebScene
  * Spec - https://developers.arcgis.com/web-scene-specification/objects/webscene/
  */
-export const ArcGisWebSceneLoader: LoaderWithParser = {
+export const ArcGisWebSceneLoader: LoaderWithParser<ArcGisWebSceneData, never, LoaderOptions> = {
```
Comment (ibgreen) [inline modules/i3s/src/arcgis-webscene-loader.ts:14]: - We should define and export a unique LoaderOptions type.

File: modules/i3s/src/i3s-building-scene-layer-loader.ts [hunk 7]
```diff
@@ -10,7 +10,7 @@ const VERSION = typeof __VERSION__ !== 'undefined' ? __VERSION__ : 'beta';
 /**
  * Loader for I3S - Building Scene Layer
  */
-export const I3SBuildingSceneLayerLoader: LoaderWithParser = {
+export const I3SBuildingSceneLayerLoader: LoaderWithParser<BuildingSceneLayerTileset, never, LoaderOptions> = {
```
Comment (ibgreen) [inline modules/i3s/src/i3s-building-scene-layer-loader.ts:13]: Unique options type!

File: modules/i3s/src/i3s-content-loader.ts [hunk 8]
```diff
@@ -10,7 +10,7 @@ const VERSION = typeof __VERSION__ !== 'undefined' ? __VERSION__ : 'beta';
 /**
  * Loader for I3S - Indexed 3D Scene Layer
  */
-export const I3SContentLoader: LoaderWithParser = {
+export const I3SContentLoader: LoaderWithParser<I3STileContent, never, LoaderOptions> = {
```
Comment (ibgreen) [inline modules/i3s/src/i3s-content-loader.ts:13]: Unique options type!

File: modules/i3s/src/i3s-loader.ts [hunk 9]
```diff
@@ -22,7 +22,7 @@ export type I3SLoaderOptions = LoaderOptions & {
 /**
  * Loader for I3S - Indexed 3D Scene Layer
  */
-export const I3SLoader: LoaderWithParser = {
+export const I3SLoader: LoaderWithParser<unknown, never, LoaderOptions> = {
```
Comment (ibgreen) [inline modules/i3s/src/i3s-loader.ts:25]: Ditto

File: modules/i3s/src/i3s-loader.ts [hunk 10]
```diff
@@ -47,7 +48,7 @@ export const I3SLoader: LoaderWithParser = {
   }
 };
 
-async function parseI3S(data, options: I3SLoaderOptions = {}, context) {
+async function parseI3S(data, options: I3SLoaderOptions = {}, context): Promise<I3STilesetHeader> {
```
Comment (belom88) [inline modules/i3s/src/i3s-loader.ts:51]:
```suggestion
async function parseI3S(data, options: I3SLoaderOptions = {}, context): Promise<I3STilesetHeader | I3STileHeader | I3STileContent> {
```

File: modules/i3s/src/i3s-loader.ts [hunk 11]
```diff
@@ -21,7 +22,7 @@ export type I3SLoaderOptions = LoaderOptions & {
 /**
  * Loader for I3S - Indexed 3D Scene Layer
  */
-export const I3SLoader: LoaderWithParser = {
+export const I3SLoader: LoaderWithParser<I3STilesetHeader, never, LoaderOptions> = {
```
Comment (belom88) [inline modules/i3s/src/i3s-loader.ts:25]:
```suggestion
export const I3SLoader: LoaderWithParser<I3STilesetHeader | I3STileHeader | I3STileContent, never, LoaderOptions> = {
```

File: modules/i3s/src/lib/parsers/parse-i3s.ts [hunk 12]
```diff
@@ -82,10 +82,10 @@ export function normalizeTileNonUrlData(tile : I3SMinimalNodeData): I3STileHeade
   return {...tile, mbs, boundingVolume, lodMetricType, lodMetricValue, type, refine};
 }
 
-export async function normalizeTilesetData(tileset : SceneLayer3D, options : LoaderOptions, context: LoaderContext): Promise<I3STilesetHeader> {
+export async function normalizeTilesetData(tileset : SceneLayer3D, options : LoaderOptions, context: LoaderContext): Promise<I3STileHeader | I3STilesetHeader> {
   const url = context.url;
   let nodePagesTile: I3SNodePagesTiles | undefined;
-  let root: I3STileHeader;
+  let root: I3STileHeader | I3STilesetHeader;
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-i3s.ts:88]: This doesn't look correct

**Pull request merged**